### ISSUE TITLE
Treat all builtin FunctionDecls as definitions [phab:D1131]

### ISSUE THREAD

**Issue by kythe** [google/kythe#1489]
_Created by **salguarnieri** at 2016-11-02 00:47:36_: 

A "builtin" node may have a declaration and a definition. Since we
create node IDs for builtin functions in a special way, we will get
an error if we record a declaration (incomplete) and a definition
(complete). To fix this, we treat all builtin functiondecls as definitions.

**Comment by kythe** [comment]
_**salguarnieri** wrote at 2016-11-02 00:52:36_:

I'm really unsure about this, but figured this was a good way to start the discussion about what to do.

An example of the offending code is the function isalnum.

**Comment by kythe** [comment]
_**zarko** wrote at 2016-11-02 02:19:58_:

If possible, we should always treat builtins as definitions (even if their definition sites are missing and even if they appear as declarations) so that any reference to a declaration of a builtin is a reference to the same node. (`isalnum` might also have a definition written out in code, but if it's not the one Clang's using we shouldn't link to it.) I'd also check on offenders like `memcpy` which probably have special-case lowering.

I believe the comment about unstable declaration locations comes from Clang's behavior of summoning the declaration syntax out of nothing when it's required; this locates the builtin at an arbitrary place that depends on things like the number of macro expansions that have been performed.

**Comment by kythe** [comment]
_**salguarnieri** wrote at 2016-11-02 16:36:08_:

  - Treat all builtin func decls as definitions.

**Comment by kythe** [comment]
_**salguarnieri** wrote at 2016-11-02 16:36:14_:

Updated to treat all bulitins as definitions. I'm a little concerned since that means we visit two versions of this functiondecl. They both refer to the same chunk of code, so I suppose they shouldn't attach contradicting information to the node.

**Comment by kythe** [comment]
_**zarko** wrote at 2016-11-02 23:26:03_:

Is there a reduced test case that can trigger the behavior where the implicit decl/def are both visited? In this c99 test I only see one visit for `isalnum`.

```
//- isalnum ref BuiltinFn
//- ImplicitLoc=vname("isalnum#n#builtinsyntactic", "", "", "", "c++")
//-   defines/binding BuiltinFn
//- !{ ImplicitLoc.loc/start _
//-    ImplicitLoc.loc/end _ }
int main() { isalnum(1); }
```

**Comment by kythe** [comment]
_**zarko** wrote at 2016-11-02 23:53:46_:

We have to be careful to avoid generating *different* names for builtins or the definition sites for builtins (otherwise we get lots and lots of redundant data; This Actually Happened). The c99 test checks for an explicit signature on the anchor (normally a bad thing, but here we really want to make sure it's stable). It should also check that BuiltinFn is `(isalnum#n#builtin, "", "", "", c++)`.

**Comment by kythe** [comment]
_**salguarnieri** wrote at 2016-11-03 00:16:27_:

  - Duplicate builtin node test case

**Comment by kythe** [comment]
_**salguarnieri** wrote at 2016-11-03 00:47:10_:

  - Make the testcase as small as possible

**Comment by kythe** [comment]
_**zarko** wrote at 2016-11-03 00:49:35_:

Works on my machines.

**Comment by kythe** [comment]
_**zarko** accepted this revision at 2016-11-03 00:49:35_

**Comment by kythe** [comment]
_**salguarnieri** closed via commit d70adf5867a686e4fee6b5a7dfcc3c3fa47bcb16 at 2016-11-03 00:52:21_
